fixes #24008 - add graphql mutations

timogoebel opened the pull request:
This is part five of the graphql series.
It adds basic mutations (create, update, delete) for models.

---
Other parts:
Part 1 - JWT Auth: #5596
Part 2 - Graphql Scaffolding: #5680
Part 3 - Relay Global ID #5681
Part 4 - Connections with totalCount #5733
Part 5 - Basic mutations for model #5736

timogoebel committed:
fixes #24008 - add graphql mutations

dariuszb-iRonin commented on app/graphql/input_types/model_attributes.rb:
```diff
@@ -0,0 +1,10 @@
+module InputTypes
+  class ModelAttributes < InputTypes::BaseInputObject
```
This is unused currently. I guess you wanted to reuse that in create and update mutations. I tried to do that by myself, however I wasn't able to find correct way of reusing it in mutations.

dariuszb-iRonin commented on app/graphql/mutations/update_model.rb:
```diff
@@ -0,0 +1,45 @@
+module Mutations
+  class UpdateModel < BaseMutation
+    description 'Creates a new hardware model.'
```
Should be `Updates existing hardware model.`

dariuszb-iRonin commented on app/graphql/mutations/create_model.rb:
```diff
@@ -0,0 +1,42 @@
+module Mutations
+  class CreateModel < BaseMutation
```
I suggest to put all mutations related with single resource in separate namespace(`models`). It's more readable IMO.

dariuszb-iRonin commented on test/graphql/mutations/delete_model_mutation_test.rb:
```diff
@@ -0,0 +1,56 @@
+module Mutations
+  class DeleteModelMutationTest < ActiveSupport::TestCase
+    let(:model) { FactoryBot.create(:model) }
+    let(:model_id) { Foreman::GlobalId.for(model.id) }
```
`Foreman::GlobalId.for` expects to receive object instead of id

dariuszb-iRonin commented on test/graphql/mutations/update_model_mutation_test.rb:
```diff
@@ -0,0 +1,83 @@
+require 'test_helper'
+
+module Mutations
+  class CreateModelMutationTest < ActiveSupport::TestCase
```
typo, should be `UpdateModelMutationTest`

dariuszb-iRonin reviewed (commented):
@timogoebel I added few comments. I also took a deep look into the code and I think we can dry it up by creating base classes for `create`, `update` and `delete` mutations. I based on your code and created PR in my fork, so you can take a look about the solution: https://github.com/dariuszb-iRonin/foreman/pull/2.

timogoebel commented:
Rebased. The suggestion to dry-up the code still applies, see https://github.com/dariuszb-iRonin/foreman/pull/2/commits/7a6db41fd5025ffed2c4a994fd585298d4cb5d1a.

xprazak2 commented on app/graphql/mutations/base_mutation.rb:
```diff
@@ -0,0 +1,21 @@
+      permission_name = resource.permission_name(action)
+
+      unless authorizer.can?(permission_name, resource)
+        raise GraphQL::ExecutionError.new(
```
![unauthorized](https://user-images.githubusercontent.com/2664090/54608216-8a524f80-4a50-11e9-8553-4d8a85f0834c.png)

Works for now, though we should probably think about how to show errors in nice and consistent way in the future. Some errors are displayed in `data.${mutationName}.errors` while others at top level, which gives us different shape of responses. We might consider including all errors in schema as some people [suggest](https://itnext.io/the-definitive-guide-to-handling-graphql-errors-e0c58b52b5e1).

xprazak2 commented on app/graphql/mutations/delete_mutation.rb:
```diff
@@ -0,0 +1,35 @@
+
+        unless object.destroyed?
+          raise GraphQL::ExecutionError.new(
+            _('Could not destroy the record.')
```
Is there a way how to show nice errors telling us more about why we failed to destroy?

![destroy-fail](https://user-images.githubusercontent.com/2664090/54606371-f4b4c100-4a4b-11e9-846d-b43d21f42559.png)

xprazak2 commented on app/graphql/mutations/update_mutation.rb:
```diff
@@ -0,0 +1,31 @@
+    def resolve(params)
+      object = ForemanGraphqlSchema.object_from_id(params[:id], context)
+
+      object.assign_attributes(params.except(:id))
```
Should we `authorize!` and `validate_object` before we try to assign any attributes?

xprazak2 commented on app/graphql/mutations/update_mutation.rb:
```diff
@@ -0,0 +1,31 @@
+        errors = if object.save
+                   []
+                 else
+                   object.errors.map do |attribute, message|
```
Maybe extract to stay DRY as both `UpdateMutation` and `CreateMutation` use same error handling.

timogoebel replied:
Extracted to `save_object` method.

xprazak2 commented on app/graphql/types/user_error.rb:
```diff
@@ -0,0 +1,11 @@
+module Types
+  class UserError < BaseObject
```
I find the class name a bit misleading, only after reading the description it was clear that the error is not related to `User` model. How about ValidationError, AttributeError or something similar?

xprazak2 commented on app/graphql/types/user_error.rb:
```diff
@@ -0,0 +1,11 @@
+    field :message, String, null: false,
+      description: 'A description of the error'
+
+    field :path, [String], null: true,
```
Not sure if having just attribute name would be enough. On the other hand, this could be useful for nested attributes if we need them, so I think we can keep it as it is for now.

xprazak2 commented:
Nice! Based on my initial testing, everything seems to work. I left a couple of comments, mostly cosmetic stuff.

timogoebel commented:
@xprazak2: Thanks for your comments. I continued to work on this PR.

- `required?` is now automatically detected
- `resource_class` is now automatically detected
- a `DeleteMutation` now shows proper error messages
- and I addressed your comment

xprazak2 commented on app/graphql/mutations/delete_mutation.rb:
```diff
@@ -0,0 +1,26 @@
+        errors = if object.destroy
+                   []
+                 else
+                   map_errors_to_path(resource)
```
I think this should be `object` as `resource` is not defined. But when I change it, I get:

```
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on name Blueberry
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on info berry hw model updated
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on vendor_class Awesome vendor
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on hardware_model 4
2019-04-01T08:15:45 [E|app|a4efced6] You may not destroy Blueberry as it is in use!
2019-04-01T08:15:45 [I|app|a4efced6] Current user set to admin (admin)
2019-04-01T08:15:45 [I|app|a4efced6] Completed 500 Internal Server Error in 49ms (ActiveRecord: 8.7ms)
 | 
 | 
2019-04-01T08:15:45 [F|app|a4efced6]   
2019-04-01T08:15:45 [F|app|a4efced6] NoMethodError (undefined method `object' for #<Hash:0x007fbc448f25e8>
 | Did you mean?  object_id):
2019-04-01T08:15:45 [F|app|a4efced6]   
2019-04-01T08:15:45 [F|app|a4efced6] app/controllers/api/graphql_controller.rb:47:in `execute_single_graphql_query'
 | app/controllers/api/graphql_controller.rb:18:in `execute'
 | app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
 | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 | lib/foreman/middleware/telemetry.rb:10:in `call'
 | lib/foreman/middleware/catch_json_parse_errors.rb:9:in `call'
 | lib/foreman/middleware/logging_context_session.rb:22:in `call'
 | lib/foreman/middleware/logging_context_request.rb:11:in `call'
 | /home/vagrant/projects/katello/lib/katello/prevent_json_parsing.rb:12:in `call'

```

xprazak2 commented:
Awesome work, just one problem when handling errors on delete.

timogoebel commented:
> Awesome work, just one problem when handling errors on delete.

@xprazak2: Thanks, updated the code and added `graphql_name` to the mutations as it's not properly set automatically.

xprazak2 reviewed (dismissed):
Works smoothly, test failure unrelated

timogoebel commented:
@xprazak2: Thanks for reviewing and testing! If we can get this into 1.22, we'd have a nice basis that plugins can work with. :-)

xprazak2 commented:
Sure, could you rebase pls?

timogoebel commented:
@xprazak2: Rebased. I used the chance to dry up the required attribute detection by introducing a small service class to do that. Also added a bit of test coverage for this.

timogoebel commented:
Test failures are unrelated.

xprazak2 commented:
Thanks!
